src/main.cpp

```diff
@@ -0,0 +1,1605 @@
+    family_indices = {0};
+    surface_properties = {0};
+    format = {};
+    present_mode = {};
+    extent = {};
+    swapchain = {};
+    chain_create_info = {};
+
+    std::uint32_t family_index_array[2] = {
+        family_indices.drawing_family,
+        family_indices.presentation_family,
+    };
+
+    surface_properties = find_surface_properties(device, surface);
+    format = pick_swap_surface_format(surface_properties.formats);
+    present_mode = pick_swap_present_mode(surface_properties.present_modes);
+    extent = pick_swap_extent(surface_properties.capabilities, window);
+    family_indices = find_queue_family_indices(device, surface);
+
+    sharing_mode = VK_SHARING_MODE_EXCLUSIVE;
+    family_index_count = 0;
+    family_index_array_base = nullptr;
+    condition =
+        family_indices.drawing_family !=
+        family_indices.presentation_family;
+    if (condition) {
+        // If drawing and presentation have separate queues, then using
+        // concurrent mode to avoid writing image ownership transfer code.
+        sharing_mode = VK_SHARING_MODE_CONCURRENT;
+        family_index_count = 2;
+        family_index_array_base = family_index_array;
+    }
+
+    image_count = surface_properties.capabilities.minImageCount + 1;
+
+    // 0 maxImageCount stands for 'no maximum'
+    condition =
+        surface_properties.capabilities.maxImageCount > 0 &&
+        image_count > surface_properties.capabilities.maxImageCount;
+    if (condition) {
+        image_count = surface_properties.capabilities.maxImageCount;
+    }
+
+    chain_create_info.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
+    chain_create_info.surface = surface;
+    chain_create_info.minImageCount = image_count;
+    chain_create_info.imageFormat = format.format;
+    chain_create_info.imageColorSpace = format.colorSpace;
+    chain_create_info.imageExtent = extent;
+    chain_create_info.imageArrayLayers = 1;
+    chain_create_info.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
+    chain_create_info.imageSharingMode = sharing_mode;
+    chain_create_info.queueFamilyIndexCount = family_index_count;
+    chain_create_info.pQueueFamilyIndices = family_index_array_base;
+
+    // Specifying currentTransform in order to get no transform
+    chain_create_info.preTransform =
+        surface_properties.capabilities.currentTransform;
+    chain_create_info.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
+    chain_create_info.presentMode = present_mode;
+    chain_create_info.clipped = VK_TRUE;
+    chain_create_info.oldSwapchain = VK_NULL_HANDLE;
+
+    result = vkCreateSwapchainKHR(
+            logical_device,
+            &chain_create_info,
+            nullptr,
+            &swapchain);
+    if (VK_SUCCESS != result) {
+        g_msg_temp = "Failed to create swapchain: ";
+        g_msg_temp += result;
+        throw std::runtime_error(g_msg_temp);
+    }
+
+    swapchain_image_format = format.format;
+    swapchain_extent = extent;
```

Comment:
It looks like these assignments are not used in any way afterwards. Can they be removed?